File: Algorithm/2023.12/PakaOxO/징검다리.js
/**
 * BaekJoon_11561, 징검다리
 *  - 문제 분류: 이분 탐색
 */
const solution = () => {
  /* 변수 관리 */
  const input = require("fs").readFileSync("./dev/stdin/11561.txt").toString().split("\n");
  const T = +input[0];
  let answer = [];

  /* 메인 로직 */
  for (let tc = 1; tc <= T; tc++) {
    const N = +input[tc];
    const max = binarySearch(N);
    answer.push(max.toString());
  }

  /* 정답 반환 */
  return answer.join("\n");

  // 이진탐색
  function binarySearch(n) {
    let left = 1;
    let right = n;

    while (left <= right) {
      const mid = Math.floor((left + right) / 2);
      if (check(mid, n)) {
        left = mid + 1;
      } else {
        right = mid - 1;
      }
    }

    return right;
  }

  // 해당 개수만큼 밟을 수 있는지
  function check(target, n) {
    let [left, right] = [1, target];

    while (left <= right) {
      const mid = Math.floor((left + right) / 2);
      if ((mid * (mid + 1)) / 2 <= n) {
        left = mid + 1;
      } else {
        right = mid - 1;
      }
    }

    return right >= target;
  }
};

console.log(solution());

File: Algorithm/2023.12/wolfy916/알고리즘 테마정리/이진 탐색/징검다리.py
'''
징검다리 - 백준 11561, 실버3
분류 : 이진 탐색
'''
import sys

# [A] 입력함수 초기화
def input():
    return sys.stdin.readline().rstrip('\n')

# [B] 1부터 n까지의 합
def Gauss_sum(n):
    return (1 + n) * n // 2

# [C] 이진 탐색 중간값 유효성 검사 함수
def check(N, mid):
    return True if N >= Gauss_sum(mid) else False

# [D] 이진탐색 함수
def binary_search(N):
    answer = 0
    left, right = 1, 10**16
    while left <= right:
        mid = (left + right) // 2
        if check(N, mid):
            left = mid + 1
            answer = mid
        else:
            right = mid - 1
    return answer

# [main]
if __name__ == '__main__':
    T = int(input())
    for _ in range(T):
        N = int(input())
        print(binary_search(N))